critwall/src/main/java/org/opencadc/critwall/BucketSelector.java

```diff
@@ -67,18 +67,88 @@
 
 package org.opencadc.critwall;
 
+import ca.nrc.cadc.util.StringUtil;
+import java.util.Iterator;
+import java.util.TreeSet;
 import org.apache.log4j.Logger;
+import org.opencadc.inventory.Artifact;
+import org.opencadc.inventory.InventoryUtil;
+
 
-/**
- *
- * @author pdowler
- */
 public class BucketSelector {
-    private final Logger log = Logger.getLogger(BucketSelector.class);
-    public final String bucketSelector;
+    private static final Logger log = Logger.getLogger(BucketSelector.class);
+
+    private static final int MAX_PREFIX_LENGTH = 1;
+    private final String rangeMin;
+    private final String rangeMax;
+    private final TreeSet<String> bucketList = new TreeSet<String>();
+
+    public Iterator<String> getBucketIterator() {
+        return bucketList.iterator();
+    }
+
+    public BucketSelector(String selector) {
+        InventoryUtil.assertNotNull(BucketSelector.class, "selectors", selector);
+
+        String[] minMax = selector.split("-");
+        StringBuilder errMsg = new StringBuilder();
+
+        if (minMax.length > 2) {
+            throw new IllegalArgumentException("invalid bucket selector: single value or range only: "
+                + selector);
+        }
+
+        // trim and convert to lower case for consistent processing
+        rangeMin = minMax[0].trim().toLowerCase();
 
-    BucketSelector(String selectorPrefix) {
-        this.bucketSelector = selectorPrefix;
-        throw new UnsupportedOperationException("TODO");
+        if (minMax.length == 1) {
+            rangeMax = rangeMin;
+        } else {
+            rangeMax = minMax[1].trim().toLowerCase();
+        }
+
+        if (!StringUtil.hasLength(rangeMin)) {
+            errMsg.append("empty min range value: " + rangeMin + "\n");
+        }
+        if (!StringUtil.hasLength(rangeMax)) {
+            errMsg.append("empty max range value: " + rangeMax + "\n");
+        }
+
+        if (rangeMin.length() > MAX_PREFIX_LENGTH) {
+            errMsg.append("min range value greater than maxlen (" + MAX_PREFIX_LENGTH + "): " + rangeMin + "\n");
+        }
+        if (rangeMax.length() > MAX_PREFIX_LENGTH) {
+            errMsg.append("max range value greater than maxlen (" + MAX_PREFIX_LENGTH + "): " + rangeMax + "\n");
+        }
+
```

Comment:
- last review said "where each section completes or throws an exception"; otherwise, down on line 127 you are still looking in the Artifact.URI_BUCKET_CHARS string with invalid input

- the 4 `if` statements (line 110 to 122) can be reduced to

```
if (rangeMin.length() != MAX_PREFIX_LENGTH || rangeMax.length() != MAX_PREFIX_LENGTH) {
    throw new IllegalArgumentException("invalid config: ....");
}
```
If you want to make it more clear that in principle that is a max and not the only length, it could be
```
if ((1 <= rangeMin.length() && rangeMin.length() <= MAX_PREFIX_LENGTH)
    && (1 <= rangeMax.length() && rangeMax.length() <= MAX_PREFIX_LENGTH) {
    log.debug("acceptable length: " + rangeMin + " - " + rangeMax);
} else {
    throw new IllegalArgumentException("invalid config: ....");
}
```

Reply:
- compressed those lines, added more general check
- exceptions thrown rather than compiling into StringBuilder